src/System.Memory/tests/MemoryPool/MemoryPool.cs

```diff
@@ -0,0 +1,239 @@
+                            }
+                        }
+                    }
+                }
+                priorBlocks.Add(newBlock);
+            }
+
+            foreach (OwnedMemory<int> prior in priorBlocks)
+            {
+                prior.Dispose();
+            }
+        }
+
+        [Fact]
+        public static void RentWithDefaultSize()
+        {
+            using (OwnedMemory<int> block = MemoryPool<int>.Shared.Rent(minBufferSize: -1))
+            {
+                Assert.True(block.Length >= 1);
+            }
+        }
+
+        [Theory]
+        [MemberData(nameof(BadSizes))]
+        public static void RentBadSizes(int badSize)
+        {
+            MemoryPool<int> pool = MemoryPool<int>.Shared;
+            Assert.Throws<ArgumentOutOfRangeException>(() => pool.Rent(minBufferSize: badSize));
+        }
+
+        public static IEnumerable<object[]> BadSizes
+        {
+            get
+            {
+                yield return new object[] { 0 };
+                yield return new object[] { -2 };
+                yield return new object[] { int.MinValue };
+                yield return new object[] { int.MaxValue };
+                yield return new object[] { MemoryPool<int>.Shared.MaxBufferSize + 1 };
+            }
+        }
+
+        [Fact]
+        public static void MemoryPoolTryGetArray()
+        {
+            using (OwnedMemory<int> block = MemoryPool<int>.Shared.Rent(42))
+            {
+                Memory<int> memory = block.Memory;
+                bool success = memory.TryGetArray(out ArraySegment<int> arraySegment);
+                Assert.True(success);
+                Assert.Equal(block.Length, arraySegment.Count);
+                unsafe
+                {
+                    void* pSpan = Unsafe.AsPointer(ref MemoryMarshal.GetReference(block.Span));
+                    fixed (int* pArray = arraySegment.Array)
+                    {
+                        Assert.Equal((IntPtr)pSpan, (IntPtr)pArray);
+                    }
+                }
+            }
+        }
+
+        [Fact]
+        public static void RefCounting()
+        {
+            using (OwnedMemory<int> block = MemoryPool<int>.Shared.Rent(42))
+            {
+                block.Retain();
+                block.Retain();
+                block.Retain();
+
+                bool moreToGo;
+                moreToGo = block.Release();
+                Assert.True(moreToGo);
+
+                moreToGo = block.Release();
+                Assert.True(moreToGo);
+
+                moreToGo = block.Release();
+                Assert.False(moreToGo);
+
+                Assert.Throws<InvalidOperationException>(() => block.Release());
+            }
+        }
+
+        [Fact]
+        public static void IsDisposed()
+        {
+            OwnedMemory<int> block = MemoryPool<int>.Shared.Rent(42);
+            Assert.False(block.IsDisposed);
+            block.Dispose();
+            Assert.True(block.IsDisposed);
```

Comment:
Should we add another block.Dispose() and check that block.IsDisposed remains true?